[WIP] enh/793: add read-write lock to yamuxstream

codemaestro64 opened the pull request:
## What was wrong?

Issue #793: 

Concurrent read and write operations on the same `Yamux` stream can interleave, leading to potential data corruption and race conditions.  
The `YamuxStream` class currently lacks a dedicated concurrency lock to ensure safe concurrent read and write access.

## How was it fixed?

As the first step toward resolving this, the `ReadWriteLock` implementation was abstracted into a reusable utility module (`stream_muxer/rw_lock.py`).  
This allows it to be cleanly reused across different parts of the codebase, including `yamux` and potentially other stream muxers.

### To-Do

- [x] Abstract `ReadWriteLock` into a reusable module (`utils/rw_lock.py`)  
- [ ] Integrate `ReadWriteLock` into `YamuxStream` to prevent concurrent read/write corruption  
- [ ] Clean up commit history  

closes #793

seetadev commented:
@codemaestro64 : Wish if you could study the implementation in mplex stream first: please visit https://github.com/libp2p/py-libp2p/pull/748. Then start the implementation in yamux stream.

CCing @acul71, who is guiding you on this PR. Adding @Jineshbansal and @lla-dane, who could share reviews as you proceed with the PR.

codemaestro64 commented:
> @codemaestro64, your approach follows the same path Go libp2p took. Go had similar concurrency issues with yamux and initially solved them with read-write locks before optimizing with atomic operations.
> 
> ## Go Implementation Evolution
> 1. **Initial**: Only window locks (like current Python yamux)
> 2. **Added**: Read/write locks (like your PR) - fixed race conditions but caused performance issues
> 3. **Optimized**: Removed locks, used Go atomic operations (go-yamux v2.1.1)
> 
> ## Recommendation
> 1. **Fix the bug**: Add `()` to `ReadWriteLock()`
> 2. **Fix lint - type check issues** : ping if you need help
> 3. **Monitor performance**: Benchmark vs current implementation
> 4. **Future optimization**: Research lock-free patterns for Python async

Thanks for the detailed feedback, @acul71! I’ve addressed the following from your recommendations:

1. Fixed the bug by adding `()` to `ReadWriteLock()`.
2. Resolved the lint and type check issues.
3. Ran initial benchmarks to compare performance with the current implementation—atomic operations are indeed more performant and fine-grained than read/write locks, as you noted.

Should I proceed with implementing atomic operations in this PR?

codemaestro64 commented:
@acul71 review [here]( https://gitlab.com/acul711/py-libp2p/-/tree/enhancement/yamuxstream-lock/acul71-PR-Reviews?ref_type=heads)

seetadev commented:
@codemaestro64 : Please also add a test suite and newsfragment. Looking forward to final review soon.

acul71 commented on tests/conftest.py:
```diff
@@ -1,6 +1,26 @@
     return None
+
+
+def pytest_configure(config):
```
?

acul71 replied:
Great
